### DISCUSSION TITLE
Pictures for ReVanced Patches

### DISCUSSION THREAD

**Discussion by oSumAtrIX** [ReVanced/revanced-discussions#997] [General]
# Conclusion so far

- Hosting links is not viable due to needing a CDN, requiring the user to have internet, and needing to leak IPs and anyone who also creates their own patches would have to deal with this; pictures have to be embedded in the ReVanced Patches files

# Open questions

- [ ] Pictures for patches mean that an annotation has to be made in ReVanced Patcher, but why would ReVanced Patcher support pictures? It doesn't belong there since neither ReVanced Patches nor ReVanced CLI need it for their purposes.
- [ ] Embedding files would bloat the repository; the only way to overcome this is Git Large File Storage

> PS: Once a comment is resolved, it is marked as resolved to make room for other comments.

RE: https://discord.com/channels/952946952348270622/953965039105232906/1134158250867183806

**Comment by decipher3114** [comment]
My take on this:

Create a new GH repo as GH pages accessible from revanced.app as well as the Manager.

We can put each patch on particular pages, e.g., `baseurl.com/patchname` (because common patches exists).

The maintainer add an image if needed, as use jekyll and Markdown to create a info page (following provided predefined template).

Then, in the manager when someone clicks on this info button.
![image](https://github.com/ReVanced/revanced-discussions/assets/84389932/fce3e075-f7b5-43af-9fed-e90af78391a1)
The page is shown in the webview with smaller window (if possible) or we could do it fullscreen(no issue in that too).

Maintainer will fork patches + Integrations + This new Repo and maintains them.

As @Ushie said, The maintainer finalises the patch, takes the screenshot and push this in their respective places.

**Reply by Ushie** [reply]
I already said that screenshots can be directly in the patches as an annotation containing a data URi, there's no need for an overcomplicated repository containing a bunch of images

**Comment by decipher3114** [comment]
Now thats rude, why my comment is hidden??
is it inappropriate?

**Reply by oSumAtrIX** [reply]
It was marked as resolved:

![image](https://github.com/ReVanced/revanced-discussions/assets/13122796/3b8e1e46-1f2e-4db9-a4d5-c32cb3023644)

This is a discussion, please don't spam it with offtopic content.

**Comment by alexandreteles** [comment]
We have an open issue for the CDN: https://github.com/ReVanced/revanced-cdn-action/issues/15

It's attributed to @PalmDevs and @reisxd (the JS devs kek). The infrastructure side of this issue is already done. We only need to finish the CI/CD aspect of it.

**Reply by oSumAtrIX** [reply]
The following has been concluded already:

> Hosting links is not viable due to [...] requiring the user to have internet, and needing to leak IPs and anyone who also creates their own patches would have to deal with this; pictures have to be embedded in the ReVanced Patches files

**Reply by alexandreteles** [reply]
The need for a CDN would be one less issue; the others would remain. Leaking IPs can be solved by proxying requests via our API, and I wonder if embedding the image files wouldn't result in considerably larger bundles.

**Reply by oSumAtrIX** [reply]
Yes, the need for a CDN is less of an issue, but its other problems would remain, such as leaking the IP or having others that provide patches themselves host a CDN.

> I wonder if embedding the image files wouldn't result in considerably larger bundles.

This is currently an "Open question" in the discussion description:

>  Embedding files would bloat the repository; the only way to overcome this is Git Large File Storage

**Reply by alexandreteles** [reply]
LFS is quite costly on GitHub. We would not pass the free 1GB tier fast enough for it to be an issue, but it should be considered before making this decision. Let's say we decide on using LFS; which image formats would we accept, and which requirements would be in place for PRs that include images to be accepted?

**Reply by oSumAtrIX** [reply]
Implementation details, such as which format we chose for the images, would be something for later.
If LFS is not a solution, the only way to embed images is by adding them to the repository. as assets. With +140 patches and approximately 100 images, the file could grow, maybe considerably, but it does not excuse the issues that a CDN would bring with (such as having to host it, it being "decoupled" from the patches, leaking IP and so on)

**Comment by TheAabedKhan** [comment]
Can we not encode the pictures into base64 and then add a `picture` key with every patch in the json file and put the encoded string there? And we can decode the image in the app. So, once the patches are shown, the images are shown instantly as well. Extra bandwidth not needed.

**Reply by Ushie** [reply]
Yep, there can be a @Preview() annotation that takes a data URi string

**Reply by oSumAtrIX** [reply]
Pictures are treated as assets, and thus will be added as assets in the project. ReVanced Patcher will consume them and provide you with a stream you can consume to display the image, for example. Though, this is an implementation detail which is something for later
